README
libweb 

Libweb is a tiny, portable and fast library for handling HTTP requests. 
It's written in C as backbone for any tool that needs to perform GETs, POSTs and other HTTP calls.



ChangeLog
2010-05-11 Paolo Perego <thesp0nge@gmail.com>

* libweb.c (web_http_server): first implementation

2010-05-07 Paolo Perego <thesp0nge@gmail.com>

* commons.c (string2url): if the URL ends with '/' it is not copied into the resource field
* get.c (web_get): HTML response header parsing

2010-04-26 Paolo Perego <thesp0nge@gmail.com>

* log.c (debug): I finally got rid of an unwanted SEGFAULT. man ctime_r is the answer.
* README.textile: improved project information

2010-04-23 Paolo Perego <thesp0nge@gmail.com>

* test/test_get.c (none): create a test case for web_get()
* log.c (none): created a small logging facility. VERY BASIC

2010-04-16 Paolo Perego <thesp0nge@gmail.com>

* commons.c (string2url): routine is almost complete with some tests performed.
* get.c (web_get): first implementation
* commons.h (none): removed

2010-04-15 Paolo Perego <thesp0nge@gmail.com>

* test/ (none): created test directory
* commons.c (none): created file. It will contains common use utilities.

2010-04-14 Paolo Perego <thesp0nge@gmail.com>
* get.c (none): created file. It will contain GET implementation
* libweb.h (none): defining an http_response data structure
* libweb.h (none): HTTP HELO sequence defined
* libweb.c (none): added COPYING notice as header
* libweb.c (web_helo): using HELO constant instead of local variable


libweb_test.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h> // memset()
#include "config.h"
#include "libweb.h"

int main(int argc, char *argv[]) {
	int sd, l;
	char buffer[8192];
	fprintf(stdout, "%s v%s\n", PACKAGE_NAME, PACKAGE_VERSION);
	if ((sd = web_client("localhost", 80)) == -1) {
		return -1;
	}
	fprintf(stdout, "%s: sending HELO... ", PACKAGE_NAME);
	fflush(stdout);
	memset(buffer, 0, 8192);
	if ((l = web_helo(sd, buffer, 8192)) < 0)
		return -1;
	fprintf(stdout, "done.\n");
	printf("%s (%d)\n", buffer, l);
	printf("Ret code: %d\n", web_http_code(buffer));
	strcpy(buffer, "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.");
	printf("Ret code should be -1 now: %d\n", web_http_code(buffer));
	
	return 0;
}

test/test_get.c
// [The "BSD licence"]
//
// test_get.c - web_get() test program
//
// test_get is part of libweb library
//
// Copyright (c) 2010 Paolo Perego, paolo@armoredcode.com
// http://www.armoredcode.com
// http://www.linkedin.com/in/thesp0nge
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
// 3. The name of the author may not be used to endorse or promote products
//    derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
// IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
// NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#include "libweb.h"

int main(int argc, char **argv) {
	http_response *res = NULL;
	
	
	fprintf(stderr, "call web_get() over existing localhost root...");
	res = web_get("http://localhost/");
	
	if (res == NULL) {
		fprintf(stderr, "received NULL form server\n");
		return -1;
	}
	fprintf(stdout, "return code is %d\n", res->code);
	if (!strcmp(res->server, NO_SERVER_HEADER))
		fprintf(stdout, "server banner is: %s\n", res->server);
	else 
		fprintf(stdout, "no server banner\n");
	fprintf(stdout, "got this from server\n%s\n", res->response);
	
	return 0;
	
}

test/test_log.c
#include <stdio.h>
#include "log.h"

int main ( void ) {
	debug("this is a debug msg\n");
	
	return 0;
} 

get.c
// [The "BSD licence"]
//
// get.c - GET HTTP method implementation
//
// Copyright (c) 2010 Paolo Perego, paolo@armoredcode.com
// http://www.armoredcode.com
// http://www.linkedin.com/in/thesp0nge
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
// 3. The name of the author may not be used to endorse or promote products
//    derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
// IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
// NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <errno.h>
#include "libweb.h"
#include "log.h"

/*
 * Sends an HTTP GET request and wait for a response.
 *
 * @param url the url we want to retrieve
 * @return the HTTP response from the web server or NULL if an error occured.
 */
http_response *web_get(char *urlstr) {
	http_response *ret;
	url *__url;
	int sd, rb, wb;
	char *buffer, *get_string;
	
	if (urlstr == NULL) {
		fprintf(stderr, "%s: NULL pointer dereference detected. Giving up.\n", __FUNCTION__);
		return NULL;
	}
	
	if ( (__url = string2url(urlstr)) == NULL ) {
		fprintf(stderr, "%s: invalid url (\"%s\")\n", __FUNCTION__, urlstr);
		return NULL;
	}
	
	sd = web_client(__url->website, __url->port);
	if (sd < 0) {
		fprintf(stderr, "%s: can't communicate with %s\n", __FUNCTION__, __url->website);
		return NULL;
	}
	
	get_string = malloc(512);
	memset(get_string, 0, 512);
	sprintf(get_string, "GET %s HTTP//1.0\n\n", __url->resource);
	
	if ( ( wb = web_send(sd, get_string, strlen(get_string))) <0 ) {
		fprintf(stderr, "%s: can't get %s\n", __FUNCTION__, __url->resource);
		return NULL;
	}
	
	// get at least BUFFER_RESPONSE bytes
	buffer = malloc(BUFFER_RESPONSE);
	if (buffer == NULL) {
		fprintf(stderr, "%s: aiee, a malloc() error here (%s)\n", __FUNCTION__, strerror(errno));
		return NULL;
	}
	
	memset(buffer, 0, BUFFER_RESPONSE);
	rb = web_recv(sd, buffer, BUFFER_RESPONSE, GET_TIMEOUT);
	if (rb == 0) {
		fprintf(stderr, "%s: I just read 0 bytes from remote host\n", __FUNCTION__);
		return NULL;
	}
	
	// preparing return value
	ret = malloc(sizeof(http_response));
	
	// copying raw http response
	strcpy(ret->response, buffer);
	
	// splitting http return code
	ret->code = web_http_code(buffer);
	if (ret->code == -1) {
		fprintf(stderr, "%s: aiee, invalid response received from web server\n", __FUNCTION__);
		return NULL;
	}
	
	// extracting server banner if present
	web_http_server(buffer, ret);
	return ret;
}

log.c
// [The "BSD licence"]
//
// log.c - logging facilities for libweb
//
// Copyright (c) 2010 Paolo Perego, paolo@armoredcode.com
// http://www.armoredcode.com
// http://www.linkedin.com/in/thesp0nge
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
// 3. The name of the author may not be used to endorse or promote products
//    derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
// IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
// NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <assert.h>

int debug(const char *msg) {
	time_t tloc;
	char buf[26];
	
	tloc = time(NULL);
	memset(buf, 0, 26);
	ctime_r((const time_t *)&tloc, buf);
	buf[strlen(buf)-1]='\0'; // removing trailing \n
	
	if (msg==NULL) {
		fprintf(stdout, "%s - [DEBUG]: some bug around. debug(NULL) has been called\n", buf);
		return -1;
	}
	
	return fprintf(stdout, "%s - [DEBUG]: %s\n", buf, msg);
}

libweb.h
#ifndef __LIBWEB_H

#define __LIBWEB_H
#define LIBWEB_ERROR	-1;
#define LIBWEB_TIMEOUT	-2;

#define HELO				"GET / HTTP\\1.0\n\n"
#define NO_SERVER_HEADER	"(none)"
#define SERVER_HEADER		"Server: "
#define BUFFER_RESPONSE		524288
#define BUFFER_LINE_SIZE	80

// if REJECT_NONWEB_PROTOCOLS is defined, an URI with a protocol
// different than both http then https will be rejected
#define REJECT_NONWEB_PROTOCOLS

#define HTTP_OK				200
#define HTTP_MISSING		404

#define GET_TIMEOUT			5 // seconds

typedef struct hr {
	int code;
	int rtt; // round trip time - useful for HTTP pinging
	char date[BUFFER_LINE_SIZE];
	char server[BUFFER_LINE_SIZE];
	char response[BUFFER_RESPONSE];
} http_response;

/*
	TODO: check the RFC for max website length 
*/
typedef struct uu {
	char proto[10];
	char website[512];
	int port;
	char resource[8192];
} url;
	
static int known_port[]={80, 8080, 8008, 8009, 8010};
static int last_http_code;

// high level APIs
http_response *web_get(char *urlstr);
url *string2url(char *urlstring);

// low level APIs
int web_client(char *hostname, unsigned int port);
int web_recv(int sd, void *buffer, size_t len, int to);
int web_send(int sd, void *buffer, size_t len);
int web_helo(int sd, char *buffer, size_t len);
int web_http_code(const char *buffer);
int web_http_server(const char *buffer, http_response *hr);
#endif //#ifndef __LIBWEB_H

commons.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include "libweb.h"

void __string2url(char *urlstring, url *url) {
	// p is the pointer to url 2 port delimiter, 
	// the ':' between the FQDN and the port number
	char *p;
	
	// r is the pointer to FQDN 2 webresource delimiter, 
	// the '/' character
	char *r;
	char tmp[4096];
	
	memset(tmp, 0, 4096);
	// No proto here. The url string is in the form:
	// fqdn[:port][/resource]
	
	// Since no protocol preamble is there, 
	// if a ':' or a '/' is present it will be the port 
	// and web resource delimiters
	p = strchr(urlstring, ':');
	r = strchr(urlstring, '/');
	if ( p == NULL) { 
		if ( r == NULL ) {
			// urlstring = www.website.com
			memcpy(url->website, urlstring, strlen(urlstring));
			strcpy(url->resource, "/");
		} else {
			// urlstring = www.website.com/resource
			memcpy(url->website, urlstring, strlen(urlstring) - strlen(r));
			if (strlen(r) == 1) 
				strcpy(url->resource, "/");
			else
				strcpy(url->resource, r + (1 * sizeof(char)));
		}
		url->port = 80;
	} else {
		if (r == NULL) {
			// urlstring = www.website.com:port
			memcpy(tmp, (p+1 * sizeof(char)), strlen(urlstring));
			url->port = atoi(tmp);
			strcpy(url->resource, "/");
		} else {
			// urlstring = www.website.com:port/resource
			memcpy(tmp, (p+1 * sizeof(char)), strlen(urlstring) - strlen(r));
			url->port = atoi(tmp);
			memcpy(url->resource, r + (1 * sizeof(char)), strlen(r));
		}
		memcpy(url->website, urlstring, strlen(urlstring) - strlen(p));
	}
	
	return ;
}
/*
 * Take a string and tokenize it trying to extract url information.
 * If the parameter string is not a good url, a NULL value will be returned.
 *
 * @param urlstring the string containing a url to call
 * @return a url information or NULL if an error occured.
 */
url *string2url(char *urlstring) {
	url *ret;

	char *p_string;
	
	
	char tmp[4096];
	
	if (urlstring == NULL) {
		fprintf(stderr, "%s: NULL pointer dereference detected. Giving up.\n", __FUNCTION__);
		return NULL;
	}
	if (strcmp(urlstring, "") == 0) {
		fprintf(stderr, "%s: an empty string is not a good URL\n", __FUNCTION__);
		return NULL;
	}
	
	// Ok, it seems that we've a real string to process.
	// I expect a string in the form:
	//		[proto://]fqdn[:port][/resource]
	//
	// Note that also resource is optional, and we will force
	// the website root ('/') in that case.
	// Also proto can be optional, in this case proto will be 
	// forced to http and port number to 80.
	
	ret = malloc(sizeof(url));
	if (ret == NULL) {
		fprintf(stderr, "%s: malloc error (%s)\n", __FUNCTION__,  strerror(errno));
		return NULL;
	}
	memset(ret, 0, sizeof(url));
	
	
	if ((p_string = strstr(urlstring, "://")) == NULL) {
		__string2url(urlstring, ret);
		strcpy(ret->proto, "http");
	} else {
		strcpy(tmp, p_string + 3 *sizeof(char));
		__string2url(tmp, ret);
		memcpy(ret->proto, urlstring, strlen(urlstring)-strlen(p_string));
	}
#ifdef REJECT_NONWEB_PROTOCOLS
	if (strcmp(ret->proto, "http") != 0 && strcmp(ret->proto, "https") != 0)
		return NULL;
#endif
	return ret;
}

libweb.c
// [The "BSD licence"]
//
// libweb - tiny, portable and fast HTTP communcation library
//
// Copyright (c) 2010 Paolo Perego, paolo@armoredcode.com
// http://www.armoredcode.com
// http://www.linkedin.com/in/thesp0nge
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
// 3. The name of the author may not be used to endorse or promote products
//    derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
// IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
// NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <netdb.h>
#include <string.h>
#include <sys/socket.h>

#include "libweb.h"
#include "log.h"

int web_send(int sd, void *buffer, size_t len)
{
	int nb = 0, wb = 0;

	for (nb=0; nb < len; nb+=wb) {
		wb = write(sd, (const void *)buffer, len);
		if (wb == -1) 
			return fprintf(stderr, "%s(): write error (%s)", __FUNCTION__, strerror(errno));
	}
	return nb;
}

int web_recv(int sd, void *buffer, size_t len, int to)
{
	fd_set fds;
	int n;
	struct timeval tv;

	// set up the file descriptor set
	FD_ZERO(&fds);
	FD_SET(sd, &fds);

	// set up the struct timeval for the timeout
	tv.tv_sec = to;
	tv.tv_usec = 0;

	// wait until timeout or data received
	n = select(sd+1, &fds, NULL, NULL, &tv);
	if (n == 0) return LIBWEB_TIMEOUT; // timeout!
	if (n == -1) return LIBWEB_ERROR; // error

	// data must be here, so do a normal recv()
	return recv(sd, buffer, len, 0);
}

/*
 * Creates a client communication endpoint to the specified hostname and to the
 * specified port number.
 * This routine takes care of calling:
 * 	+ socket()
 * 	+ connect()
 *
 * @param hostname the server hostname
 * @param port the server port number to connect to
 * @return a valid socket descriptor or -1 if an error occurs
 */
int web_client(char *hostname, unsigned int port) 
{
	int sd = 0;
	struct sockaddr_in sin;
	struct hostent *he;

	if ((hostname == NULL) || (port <=0)) {
		fprintf(stderr, "attempt to feed %s() with invalid parameters (%s, %d)\n", 
				__FUNCTION__, 
				hostname,
				port);
		return -1;
	}
	// 1. cook the socket...
	if ((sd = socket (PF_INET, SOCK_STREAM, 0)) == -1) {
		fprintf(stderr, "%s(): can't cook socket (%s)\n", 
				__FUNCTION__, strerror(errno));
		return -1;
	}
	
	// 2. fill my peer information
	memset(&sin, 0, sizeof(struct sockaddr_in));
	sin.sin_family = AF_INET;
	sin.sin_port = htons(port);
	// 2.1 resolve server hostname
	if (inet_aton(hostname, &(sin.sin_addr)) == 0) {
		he = gethostbyname(hostname);
		if (he == NULL) {
			fprintf(stderr,"%s(): can't resolve %s (%s)\n",
				__FUNCTION__,
				hostname,
				hstrerror(h_errno));
			return -1;
		}
		memcpy(&(sin.sin_addr), he->h_addr, sizeof(struct in_addr));

	}
	// 3. so please... connect
	if ( connect(sd, (const struct sockaddr *)&sin, sizeof(struct sockaddr)) == -1) {
		fprintf(stderr,"%s(): can't connect to %s:%d (%s)\n",
			__FUNCTION__,
			hostname,
			port,
			strerror(errno));
		return -1;
	}
	return sd;
}

/*
 * Sends a GET / HTTP/1.0 request to get web server return banner.
 * 
 * @param sd: a previously opened socked descriptor.
 * @param buf: a previously allocated memory buffer to store web server answer
 * @param len: the buffer lenght
 * @return the number of bytes read from the web server or -1 if some error occured.
 */
int web_helo(int sd, char *buffer, size_t len) {
	
	if ((sd<3)) {
		fprintf(stderr, "%s(): invalid or reserved file descriptor (%d)\n", 
			__FUNCTION__, 
			sd);
		return -1;
	}
	if (buffer == NULL) { 
		fprintf(stderr, "%s(): a NULL memory pointer has been passed as memory pointer. Don't.\n", 
			__FUNCTION__); 
		return -1;
	}
	
	if (web_send(sd, HELO, strlen(HELO))<0)
		return -1;
	
	// fprintf(stderr, "%s() a custom implementation of strlen(), must be provided", __FUNCTION__);
	
	/*
	 * This is a sample answer for the HTTP helo given from my apache installation on my 
	 * SnowLeopard MacBook.
	 */
	
	// HTTP/1.1 200 OK
	// Date: Mon, 22 Mar 2010 17:17:55 GMT
	// Server: Apache/2.2.13 (Unix) mod_ssl/2.2.13 OpenSSL/0.9.8l DAV/2
	// Content-Location: index.html.en
	// Vary: negotiate
	// TCN: choice
	// Last-Modified: Sat, 20 Nov 2004 20:16:24 GMT
	// ETag: "c39dd9-2c-3e9564c23b600"
	// Accept-Ranges: bytes
	// Content-Length: 44
	// Connection: close
	// Content-Type: text/html
	// Content-Language: en
	
	int ret = web_recv(sd, buffer, len, 5);
	
	return ret;
}
/*
 * Looks in the given buffer for an HTTP/1.1 header so to give the return code
 * 
 * @param buffer the char buffer containing, maybe, an HTTP answer
 * @return the HTTP return code or -1 if something wrong occured
 */
int web_http_code(const char *buffer) 
{
	char *p = NULL;
	char *s = NULL;
	char code_s[4];
	
	if (buffer == NULL) { 
		fprintf(stderr, "%s(): a NULL memory pointer has been passed as memory pointer. Don't.\n", 
			__FUNCTION__); 
		return -1;
	}
	p = strstr(buffer, "HTTP/1.");
	if (p == NULL) {
		fprintf(stderr, "%s(): doesn't seem to be a good HTTP response block\n", __FUNCTION__);
		return -1;
	}
	
	s = p+(9*sizeof(char));
	strncpy(code_s, s, 3);
	return atoi(code_s);
}

/*
 * Looks in the given buffer for a Server: field and copy it to the http_response parameter.
 * This information can be used for web server footprinting pourposes.
 *
 * @param buffer the char buffer containing, maybe, an HTTP answer
 * @param hr the http_response variable to be used to store results.
 * @return 0 if the Server: header has been coped to hr, or -1 otherwise
 */
int web_http_server(const char *buffer, http_response *hr) {
	char *s;
	char *nl;
	if (buffer == NULL || hr == NULL) { 
		fprintf(stderr, "%s(): a NULL memory pointer has been passed as memory pointer. Don't.\n", 
			__FUNCTION__); 
		return -1;
	}
	
	s = strstr(buffer, SERVER_HEADER);
	if (s == NULL) {
		strcpy(hr->server, NO_SERVER_HEADER);
		return -1;
	}
	// SERVE UNA STRCPY che COPY FINO AD UN CERTO CARATTERE
	nl = strstr(s+1*sizeof(char), "\n");
	strncpy(hr->server, s+(strlen(SERVER_HEADER) * sizeof(char)), strlen(s)-strlen(nl));
	
	debug(hr->server);
	printf("---\n");
	return 0;
}
